### PR TITLE
Add scrolling

### PR DESCRIPTION
**Pull request by aki237**
Just one issue at a time.... :+1:

### CODE REVIEW

**Comment by aki237** [comment]
This resolves #1 issue

**Comment by zachwhaley** [comment]
Cool, so now for the code review.

I'm going to be making line by line comments and suggestions.  If you disagree with a comment, make sure to explain yourself :)

File: src/buffer.hpp [hunk 1]
```diff
@@ -22,6 +22,7 @@ class Buffer
     int getchar() const;
     void update(int ch);
     void display();
+    unsigned int cursorY;
```
Comment (zachwhaley) [inline src/buffer.hpp]: Could you explain to me what this variable does?

Its name is so familiar to Cursor.y but I don't think it is serving the same purpose as that variable.

Also, because this variable is only ever used within the Buffer class, there is no need for it to be public.  Could you move this variable to the private member variables, ie where `m_cur`, `m_filename`, and others are.

Finally, because this is a member variable it needs a `m_` prefix. This helps distinguish member variables from free variables out in the wilds of a source file.

File: src/buffer.cpp [hunk 2]
```diff
@@ -96,17 +96,25 @@ Buffer::display()
...
         for (size_t col = 0; col < cols; col++) {
-            if (line < m_lines.size() && col < m_lines[line].length())
-                waddch(m_textwin, m_lines[line][col]);
+	    if(cursorY  > lines - 2){
+	        m_cur.y = lines - 2;
+	        scrollFactor = cursorY - lines + 2 ;
+	    }
+	    else {
+	        scrollFactor = 0;
+	    }
```
Comment (zachwhaley) [inline src/buffer.cpp]: The `if else` block above this comment needs one more space in its indentation.

To make all this indentation stuff easier, I'd suggest using one of these editors if you are not already.

Vim/[Neovim](https://neovim.io/), [Emacs](https://www.gnu.org/software/emacs/), or [Atom](https://atom.io/)

I'm pretty sure all of these editors do file type dedication and automatic indenting, although you might need to change some settings to use 4 space indentation like I have.

File: src/buffer.cpp [hunk 3]
```diff
@@ -96,17 +96,25 @@ Buffer::display()
...
     wnoutrefresh(m_titlewin);
 
     // Refresh text window
+    int scrollFactor=0;
     for (size_t line = 0; line < lines - 1; line++) {
         wmove(m_textwin, line, 0);
         for (size_t col = 0; col < cols; col++) {
-            if (line < m_lines.size() && col < m_lines[line].length())
-                waddch(m_textwin, m_lines[line][col]);
+	    if(cursorY  > lines - 2){
```
Comment (zachwhaley) [inline src/buffer.cpp]: Formatting

```c
// wrong 
if(cursorY  > lines - 2){

// right
if ( cursorY > lines - 2) {
```

File: src/buffer.hpp (hunk 1 above)
Comment (aki237) [inline src/buffer.hpp]: m_cur.(x or y) is just used to move the cursor around the screen within the
terminal width and height...... So we need another variable to controll the
printng of lines in Buffer::display() function. m_cur.y increments as we
move down so is cursorY. But m_cur.y stops incrementing when it reaches the
window height(LINES or lines) and starts to change only when cursorY is
less than LINES..... As far as naming is concerned I was just testing with
that variable and I'll change it if you need
Acually the buffer class doesn't have its own Constructor.... So when
cursorY is not initialised eli throws segfault.....

File: src/buffer.cpp [hunk 4]
```diff
@@ -96,17 +96,25 @@ Buffer::display()
...
     }
     stringstream ss;
-    ss << " " << m_filename << " " << m_cur.y << "," << m_cur.x;
+    ss << " " << m_filename << " " << cursorY << "," << m_cur.x;
     string title = ss.str();
     mvwaddstr(m_titlewin, 0, 0, title.c_str());
     wnoutrefresh(m_titlewin);
 
     // Refresh text window
+    int scrollFactor=0;
```
Comment (zachwhaley) [inline src/buffer.cpp]: Use snake case instead of camel case, and your `=` needs spaces surrounding it

```c
// wrong
int scrollFactor=0;

// right
int scroll_factor = 0;
```

File: src/buffer.cpp [hunk 5]
```diff
@@ -160,19 +171,23 @@ Buffer::prevline()
...
 Buffer::nextchar()
 {
-    if (m_cur.x != m_lines[m_cur.y].length()) {
+    if (m_cur.x != m_lines[cursorY].length()) {
         m_cur.x++;
     }
-    else if (m_cur.y != m_lines.size() - 1) {
+    else if (cursorY != m_lines.size() - 1) {
         m_cur.y++;
+	cursorY++;
```
Comment (zachwhaley) [inline src/buffer.cpp]: Indentation: Needs one more space

File: src/buffer.cpp [hunk 6]
```diff
@@ -183,8 +198,9 @@ Buffer::prevchar()
     if (m_cur.x != 0) {
         m_cur.x--;
     }
-    else if (m_cur.y != 0) {
+    else if (cursorY != 0) {
         m_cur.y--;
+	cursorY--;
```
Comment (zachwhaley) [inline src/buffer.cpp]: Indentation: Needs one more space

File: src/editor.cpp [hunk 7]
```diff
@@ -13,6 +13,7 @@ Editor::Editor()
     initscr();
     raw();
     noecho();
+    m_buffer.cursorY=0;
```
Comment (zachwhaley) [inline src/editor.cpp]: cursorY should be initialized in `buffer.hpp` where it is declared.  This is a new feature of C++11 :)

File: src/buffer.cpp [hunk 8]
```diff
@@ -96,17 +96,25 @@ Buffer::display()
...
-                waddch(m_textwin, m_lines[line][col]);
+	    if(cursorY  > lines - 2){
+	        m_cur.y = lines - 2;
+	        scrollFactor = cursorY - lines + 2 ;
+	    }
+	    else {
+	        scrollFactor = 0;
+	    }
+            if (line+scrollFactor < m_lines.size() && col < m_lines[line+scrollFactor].length())
+                waddch(m_textwin, m_lines[line+scrollFactor][col]);
```
Comment (zachwhaley) [inline src/buffer.cpp]: Need spaces surrounding `+`

```c
// wrong
line+scrollFactor

// right
line + scrollFactor
```

File: src/buffer.cpp (hunk 4 above)
Comment (aki237) [inline src/buffer.cpp]: Sorry my emacs messes up the indentation, I tried fixing in most cases, missed a few I think....

**Comment by zachwhaley** [comment]
For all of these comments, I'd suggest fixing them and then using `git commit --amend` to add these changes to your original commit.  Then use `git push -f` to rewrite this branches history.

This way the commit log stays clean and we don't get a lot of "fixed formatting" commits :)

**Comment by aki237** [comment]
Sorry my emacs messes up the indentation, I tried fixing in most cases, missed a few I think....

**Comment by aki237** [comment]
Anyway what do you want me to name cursorY??

File: src/buffer.hpp (hunk 1 above)
Comment (zachwhaley) [inline src/buffer.hpp]: Sorry for the lack of code comments that probably would have explained this (my bad), but Cursor.y was intended to represent the position of the cursor in the file, not the screen.  I can see now that that name doesn't really convey that. I think it is a good idea what you've done here, making Cursor represent the position of the cursor on the screen.  With that said, I think we should move cursorY into its own struct called Position or something that signifies that this is the position of the cursor in the file.

What do you think?

The Buffer class does have a constructor, It is just implicit.  There was nothing I needed to add to the constructor, so I didn't explicitly write it.  But regardless, we don't need it because C++11 is way cool and allows declaration and initialization within the header file :)

File: src/buffer.cpp (hunk 4 above)
Comment (zachwhaley) [inline src/buffer.cpp]: Not a problem.  Like I said before this is some good work :+1:

File: src/buffer.hpp (hunk 1 above)
Comment (aki237) [inline src/buffer.hpp]: I hear ya..... I am getting used to Cpp11 these days....... about position
that would be good.... I think ..... this is just an Idea... As the Buffer
class has many functions that modifies the cursor location, why dont we
write that into a separate class and put the functions there...

and

we dont need nextchar and prevchar instead movecursor(int moveAmount)
Similarly for line..... so that it opens up a possibility of writing
plugins😜

Comment (zachwhaley) [inline src/buffer.hpp]: Sounds interesting!  Make another pull request if you'd like :)

FYI: I was heavily influenced by the Sandy text editor, which used separate functions for moving between characters http://git.suckless.org/sandy/

**Comment by zachwhaley** [comment]
I'd create a struct named Position with y and x variables, like Cursor.  We won't use x for this change.

Then create a member variable named m_pos in Buffer of type Position.

**Comment by zachwhaley** [comment]
I pulled your code to do some testing, and you're change has some bugs in it.

While scrolling down seems to work fine, scrolling up does not.  The cursor gets stuck at the bottom of the screen until you've reached the very top of the file again.

I was going to make a couple patches to address this and ended up rewriting the scrolling code; you can see my changes here https://github.com/zachwhaley/eli/compare/scrolling

That implementation is only half done though because horizontal scrolling does not work.  If you'd like, I'll let you pull my branch, add horizontal scrolling, and create a new pull request.

**Comment by aki237** [comment]
Try changing lines - 2 to some other number

**Comment by aki237** [comment]
Yeah I knew that...  I was scrapping te whose scroll mechanism to the one similar in your branch.... it works perfect

**Comment by aki237** [comment]
In nextchar and prevchar we can call nextline and prevline respectively to move lines when out of bounds....
